src/supertux/game_session.cpp

```diff
@@ -67,31 +66,46 @@ GameSession::GameSession(const std::string& levelfile_, Savegame& savegame, Stat
   m_reset_pos(0.0f, 0.0f),
   m_newsector(),
   m_newspawnpoint(),
-  m_pastinvincibility(false),
-  m_newinvincibilityperiod(0),
+  m_invincibilitytimeleft(),
   m_best_level_statistics(statistics),
   m_savegame(savegame),
   m_play_time(0),
   m_edit_mode(false),
   m_levelintro_shown(false),
   m_coins_at_start(),
-  m_bonus_at_start(),
+  m_boni_at_start(),
   m_max_fire_bullets_at_start(),
   m_max_ice_bullets_at_start(),
   m_active(false),
   m_end_seq_started(false),
-  m_current_cutscene_text()
+  m_current_cutscene_text(),
+  m_endsequence_timer()
 {
+  m_boni_at_start.resize(InputManager::current()->get_num_users(), NO_BONUS);
+  m_max_fire_bullets_at_start.resize(InputManager::current()->get_num_users(), 0);
+  m_max_ice_bullets_at_start.resize(InputManager::current()->get_num_users(), 0);
+
   if (restart_level() != 0)
     throw std::runtime_error ("Initializing the level failed.");
 }
 
 void
 GameSession::reset_level()
 {
-  m_currentsector->get_player().set_bonus(m_bonus_at_start);
+  for (const auto& p : m_currentsector->get_players())
+  {
+    try
+    {
+      p->set_bonus(m_boni_at_start.at(p->get_id()));
+    }
+    catch(const std::out_of_range&)
+    {
```

Comment:
Can this actually happen? You're resizing the array to the number of users in the constructor.

Reply:
The number of users is dynamic; someone can plug their controller mid-game and someone else can unplug theirs. I'll try to investigate further.

Reply:
It can be reproduced: Play the game with 2 players, remove a player, press escape and restart the level: a crash will occur.